src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/parquet/shredder.h"
+
+#include "serde/parquet/schema.h"
+#include "serde/parquet/value.h"
+
+#include <seastar/coroutine/maybe_yield.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <boost/range/irange.hpp>
+
+#include <ranges>
+#include <stdexcept>
+
+using boost::irange;
+using std::ranges::reverse_view;
+
+namespace serde::parquet {
+
+namespace {
+
+// A struct to track all the level related information we need during the tree
+// shredding traversal.
+struct traversal_levels {
+    uint8_t repetition_level = 0;
+    uint8_t definition_level = 0;
+    // repetition_depth is the current number of ancestor schema nodes in the
+    // schema tree that are set to be repeated. We track this seperately because
+    // repeated is only set for the additional values in the repeated node, so
+    // the first time we repeat we need to use the current repetition_level so
+    // that assembly knows at which level the repetition is starting at, but
+    // additional repetitions need to use the current repetition_depth as the
+    // repetition_level.
+    uint8_t repetition_depth = 0;
```

Comment:
so this is used at runtime, and only the two levels above make it into any serialized representation?

Reply:
Correct! The depth tells us how many repeated parents there are so we know when the list repeats we know the level to set, as the current rep level could be anything.